packages/recommend-react/src/useSafeEffect.ts

```diff
@@ -0,0 +1,55 @@
+import { useEffect, useRef } from 'react';
+
+type Props = Record<string, any>;
+type AlteredProps<TProps> = Partial<{ [TProp in keyof TProps]: any }>;
+
+function isPrimitive(obj: any) {
+  return Object(obj) !== obj;
+}
+
+function separatePrimitives<TProps>(props: TProps) {
+  const primitives = Object.entries(props).reduce<TProps>((prev, [k, v]) => {
+    if (isPrimitive(v)) {
+      return {
+        ...prev,
+        [k]: v,
+      };
+    }
+
+    return prev;
+  }, {} as TProps);
+  const nonPrimitives = Object.entries(props).reduce<TProps>((prev, [k, v]) => {
+    if (!isPrimitive(v)) {
+      return {
+        ...prev,
+        [k]: v,
+      };
+    }
+
+    return prev;
+  }, {} as TProps);
+
+  return { primitives, nonPrimitives };
```

Comment:
```suggestion
  return Object.entries(props).reduce<{
    primitives: TProps;
    nonPrimitives: TProps;
  }>(
    (prev, [k, v]) => {
      if (isPrimitive(v)) {
        prev.primitives[k] = v;
      } else {
        prev.nonPrimitives[k] = v;
      }

      return prev;
    },
    { primitives: {} as TProps, nonPrimitives: {} as TProps }
  );
```

mutated here because local mutation in a reduce is safe